relayer/processor/path_processor.go

```diff
@@ -0,0 +1,305 @@
+				continue MsgTransferLoop
+			}
+		}
+		// Packet is not yet relayed! need to relay from src to dst
+		unrelayedPackets = append(unrelayedPackets, IBCMessageWithSequence{Sequence: transferSeq, Message: msgTransfer})
+	}
+
+	// now iterate through packet-flow-complete messages and remove any leftover messages if the MsgTransfer or MsgRecvPacket was in a previous block that we did not query
+	for ackSeq := range pathEndPacketFlowMessages.SrcMsgAcknowledgement {
+		toDeleteSrc[MsgTransfer] = append(toDeleteSrc[MsgTransfer], ackSeq)
+		toDeleteDst[MsgRecvPacket] = append(toDeleteDst[MsgRecvPacket], ackSeq)
+		toDeleteSrc[MsgAcknowledgement] = append(toDeleteSrc[MsgAcknowledgement], ackSeq)
+	}
+
+	// would iterate timeout messages here also
+
+	res.UnrelayedPackets = unrelayedPackets
+	res.UnrelayedAcknowledgements = unrelayedAcknowledgements
+	res.ToDeleteSrc = toDeleteSrc
+	res.ToDeleteDst = toDeleteDst
+}
+
+func (pp *PathProcessor) deleteCachedMessages(messages MessageCache, toDelete ...map[string][]uint64) {
+	for _, toDeleteMap := range toDelete {
+		for message, toDeleteMessages := range toDeleteMap {
+			for _, sequence := range toDeleteMessages {
+				delete(messages[message], sequence)
+			}
+		}
+	}
+}
+
+func (pp *PathProcessor) sendMessages(pathEnd *PathEndRuntime, messages []IBCMessageWithSequence) error {
+	if len(messages) == 0 {
+		return nil
+	}
+	// TODO construct MsgUpdateClient for this pathEnd, using the latest trusted IBC header from other pathEnd, prepend messages with the MsgUpdateClient, then send the messages to this pathEnd
+
+	pp.log.Debug("will send", zap.Any("messages", messages))
+
+	return nil
+}
+
+// messages from both pathEnds are needed in order to determine what needs to be relayed for a single pathEnd
+func (pp *PathProcessor) processLatestMessages(pathEnd1MessageCache MessageCache, pathEnd2MessageCache MessageCache) error {
+	var processPacketFlowWaitGroup sync.WaitGroup
+	processPacketFlowWaitGroup.Add(2)
+
+	pathEnd1PacketFlowMessages := PathEndPacketFlowMessages{
+		SrcMsgTransfer:        pathEnd1MessageCache[MsgTransfer],
+		DstMsgRecvPacket:      pathEnd2MessageCache[MsgRecvPacket],
+		SrcMsgAcknowledgement: pathEnd1MessageCache[MsgAcknowledgement],
+	}
+	pathEnd2PacketFlowMessages := PathEndPacketFlowMessages{
+		SrcMsgTransfer:        pathEnd2MessageCache[MsgTransfer],
+		DstMsgRecvPacket:      pathEnd1MessageCache[MsgRecvPacket],
+		SrcMsgAcknowledgement: pathEnd2MessageCache[MsgAcknowledgement],
+	}
+
+	// process the packet flows for both packends to determine what needs to be relayed
+	var pathEnd1ProcessRes, pathEnd2ProcessRes PathEndProcessedResponse
+	go pp.getUnrelayedPacketsAndAcksAndToDelete(pathEnd1PacketFlowMessages, &processPacketFlowWaitGroup, &pathEnd1ProcessRes)
+	go pp.getUnrelayedPacketsAndAcksAndToDelete(pathEnd2PacketFlowMessages, &processPacketFlowWaitGroup, &pathEnd2ProcessRes)
```

Comment:
Upon further reading, since `pp.getUnrelayedPacketsAndAcksToDelete` is just copying some values out of maps, I would not expect the overhead of a new goroutine to result in any speedup. I think this could be

```go
end1Res := pp.getUnrelayed...(pathEnd1PacketFlowMessages)
end2Res := pp.getUnrelayed...(pathEnd2PacketFlowMessages)
```